Modules/_threadmodule.c

```diff
@@ -1148,20 +1089,14 @@ static PyObject *
 rlock_release_save(PyObject *op, PyObject *Py_UNUSED(ignored))
 {
     rlockobject *self = (rlockobject*)op;
-    PyThread_ident_t owner;
-    unsigned long count;
 
-    if (self->rlock_count == 0) {
+    PyThread_ident_t owner = _Py_atomic_load_ullong_relaxed(&self->lock.thread);
+    size_t count = _Py_atomic_load_ulong_relaxed(&self->lock.level);
```

Comment:
I think we should just use non-atomic reads for `owner` and `count`/`level` here. The `_Py_atomic_load_ulong_relaxed` isn't correct for a `size_t` field.

We can structure this like (pseudo-code):

```c
if (!_PyRecursiveMutex_IsLockedByCurrentThread(&self->lock) {
    PyErr_SetString(...);
    return NULL;
}

PyThread_ident_t owner = self->lock.thread;
size_t count = self->lock.level; // or maybe self->lock.level + 1
_PyRecursiveMutex_Unlock(&self->lock);
return Py_BuildValue(...);
```

We don't need atomic accesses for the fields since we own the lock and those fields are only set by the lock's owner.

Reply:
fixed